trapper/metrics/metric_handlers/metric_handler.py

```diff
@@ -43,77 +43,72 @@ def __init__(
     def tokenizer(self):
         return self._tokenizer.tokenizer
 
-    @tokenizer.setter
-    def tokenizer(self, value):
-        if isinstance(value, TokenizerWrapper):
-            self._tokenizer = value
-        else:
-            raise ValueError(
-                f"The value must be an instance of a "
-                f"class derived from {TokenizerWrapper}"
-            )
-
     @property
     def label_mapper(self):
         return self._label_mapper
 
-    @label_mapper.setter
-    def label_mapper(self, value):
-        if isinstance(value, LabelMapper):
-            self._label_mapper = value
-        else:
-            raise ValueError(
-                f"The value must be an instance of a "
-                f"class derived from {LabelMapper}"
-            )
-
-    def __call__(self, dataset: datasets.Dataset) -> datasets.Dataset:
+    def __call__(self, dataset: datasets.Dataset) -> None:
         """
-        Where extraction from instances happen through the method
-        `extract_metadata()`, returns the input dataset as is. Do not override
-        this method in child class, instead use `MetricHandler.extract_metadata()`.
+        This method applies `self.extract_metadata()` to each instance of the dataset.
+        Do not override this method in child class, instead
+        override `self.extract_metadata()`.
+
+        Note:
+            This handler is only called once in trainer for each dataset. By default,
+            only eval_dataset is called.
 
         Args:
-            dataset:
+            dataset: datasets.Dataset object
 
         Returns: IndexedInstance as is.
         """
-        # Currently through HF trainer validation split is used for eval
         dataset.map(self.extract_metadata)
-        return dataset
 
     def extract_metadata(self, instance: IndexedInstance) -> None:
         """
         Child class may implement this method for metadata extraction from an instance.
+        It is designed to be read-only, i.e do not manipulate instance in any
+        way. It should return None for efficiency purposes.
 
         Args:
             instance: Current instance processed
 
         Returns: None
         """
-        pass
+        return None
 
-    def postprocess(
+    def preprocess(
```

Comment:
Lets make this method accept `pred: EvalPrediction` and return the same type, to make it more future-compatible with transformers? (Of course for child classes as well).

Reply:
jury (or in general metric) inputs are named `predictions/references`, changing this to `EvalPrediction` makes it more confusing I think, plus the name implies it like it is only predictions but the object has labels, too. This characteristic of `EvalPrediciton` requires additional knowledge of transformers library. IMHO, I disagree to change the current structure to proposed one for the reasons I mentioned.

Reply:
I think, we would just be hiding by leaving them as tuple of predictions, references instead of the `EvalPrediction` type. The reason is that, they are actually the `EvalPrediction` in disguise. If for a reason, the transformers library changed the object returned by the model.forward(), we would have to change these metrics as well. 
In fact, requiring the knowledge of `EvalPrediciton` type is fine, that is what we use already. The models in trapper returns that type (and only that type), so it is only natural to have the metrics use that type.

edit: BTW, we are implementing the features for the trapper, which is built on top of transformers and supposes complete integration. If we were talking about the Jury metric interfaces in the jury library itself, I would not oppose.